Added remote entity in addition of media player entity

albaintor opened the pull request:
Hi,

I have added support for remote entity in addition of media player.

I had to make heavy modifications in the code so a lot of tests should be done to check after regressions

albaintor commented:
I have tested it successfully : both entities work correctly on my side so far

albaintor commented on README.md:
```diff
@@ -25,6 +25,78 @@ Preview features:
+
+**Shield TV :** `SCREENSAVER`
+
+**onn. Streaming Device 4K pro:** `TELETEXT`
```
You're right, I mixed with Philips TV. I updated the readme

albaintor commented on src/config.py:
```diff
@@ -13,11 +13,30 @@
 _CFG_FILENAME = "config.json"
 
 
+def create_entity_id(avr_id: str, entity_type: EntityTypes) -> str:
+    """Create a unique entity identifier for the given receiver and entity type."""
+    return f"{entity_type.value}.{avr_id}"
```
Right, renamed

albaintor commented:
Hi, I have applied the selected changes.

I just have to test the new implementation of commands/command sequence in async mode with the delay/repeat parameters

zehnm commented on src/remote.py:
```diff
@@ -0,0 +1,150 @@
+                res = await self._device.turn_on()
+        elif command in self.options.get(Options.SIMPLE_COMMANDS, {}):
+            res = await self._device.send_media_player_command(command)
+        elif cmd_id in [Commands.SEND_CMD, Commands.SEND_CMD_SEQUENCE]:
+            _ = asyncio.get_event_loop().create_task(self.send_commands(cmd_id, params))
```
This is a nice and easy solution for `SEND_CMD_SEQUENCE`. Unfortunately this makes all simple single `SEND_CMD` best effort only, and the user gets no feedback anymore if the command couldn't be sent.

I suggest to only create a task for `SEND_CMD` if `delay > 2000 || repeat > 2`. These parameter values should be well within the client timeout and could be further adjusted if needed.

albaintor replied:
I have submitted a more precise calculation : number of commands x repeats x (delay + max exec time of a command) to decide between async or direct execution

zehnm reviewed (changes requested):
Thanks for the changes.
Only one thing left for `SEND_CMD`, then it looks good.

albaintor commented:
Just applied a math calculation to evalutate expected duration of command or command sequence

albaintor commented:
I have found a better way finally : with `asyncio.timeout` you can wait until a given timeout that cancels the command but using the `asyncio.shield` it makes the command keep running

zehnm commented:
I'll create another test release from the main branch to test the updated Python & pyinstaller versions, then come back to the remote-entities in Android & Apple TV.

zehnm commented:
It's best not to update the driver.json metadata in PRs to avoid merge conflicts. I update it when creating a release.

zehnm reviewed (approved):
Looks good, if something comes up during testing, we can always fix it later :-)

After fixing the merge conflict I'll merge it.

albaintor commented:
Hi @zehnm I made tests and I fixed several things (wrong unit for delay, parameters not correctly parsed for command sequence)

I also updated androidtvremote2 : no impacts on what we are using but a new command to send text for future use maybe

albaintor commented:
Do you think that a "true long" press using AndroidTV start/stop long will work for HDMI CEC volume (non IP) commands ?
Maybe at last we would have to ignore repeat command for those HDMI volume commands

No problem for waiting the future long wait feature anyway.

One last thing to consider with these start/stop long press : the `@retry` python macro should be applied on the start but most of all on the stop long press so that the stop gets finally processed to avoid infinite long press due to

zehnm commented:
Forgot to report back: technically it "works", but it's not usable. The longer the `LONG` command is active, the longer it continues running after sending `STOP_LONG`! 
Tested on a Shield TV with remapping the FF & rewind keys to send `START_LONG` and `STOP_LONG` in `config/profiles/NVIDIA_SHIELD.json`:
```json
{
  "command_map": {
    "fast_forward": {
      "keycode": "KEYCODE_VOLUME_DOWN",
      "action": "BEGIN"
    },
    "rewind": {
      "keycode": "KEYCODE_VOLUME_DOWN",
      "action": "END"
    }
  }
}
```
The only option I see right now is to write a custom repeat handler that sends individual commands at a certain rate, when the long-press addition becomes available in the integration-API.

zehnm commented:
The command continues to be active for a certain time after the `STOP_LONG` command (volume continues to decrease). The duration depends on how long it was active. I still need to investigate further if it's caused in the `androidtvremote2` library (e.g. by somehow delaying the command), or if it's a bug in the Android TV Remote Service itself.

zehnm commented on src/media_player.py:
```diff
@@ -0,0 +1,81 @@
+        self._device_config = device_config
+        self._profile = profile
+
+        entity_id = create_entity_id(device_config.id, EntityTypes.MEDIA_PLAYER)
```
Changing the entity identifier of the existing media-player entity is a breaking change we can't allow. Otherwise all existing user configurations won't work anymore and have to be redone (including activities, macros).

zehnm commented on src/driver.py:
```diff
@@ -222,81 +271,102 @@ async def handle_android_tv_update(atv_id: str, update: dict[str, Any]) -> None:
+    """
+    # dead simple for now: one media_player entity per device!
+    # TODO #21 support multiple zones: one media-player per zone
+    return [f"media_player.{device_id}", f"remote.{device_id}"]
```
See comment in media_player.py: existing entity may not be renamed

zehnm commented on src/driver.py:
```diff
@@ -110,88 +144,93 @@ async def on_subscribe_entities(entity_ids) -> None:
+            config.devices.update(device)
+
+        # TODO is this the correct state?
+        api.configured_entities.update_attributes(identifier, {MediaAttr.STATE: ucapi.media_player.States.STANDBY})
```
Old logic: entity states have been set already above

Sawtaytoes commented:
As an observer desperate for this fix: https://github.com/unfoldedcircle/integration-androidtv/issues/91, is there anything blocking this PR from merging?

One of the other blockers appears to be merged already: https://github.com/unfoldedcircle/integration-androidtv/pull/120, and it looks like all the checks are passing now in GitHub.

ItsHarper commented:
@Sawtaytoes I'm pretty sure the code from this PR got integrated along with the voice commands feature, even though that did not get mentioned in the release notes. Looking at the code, it does appear that long press support should be implemented correctly now.

That still leaves the question of whether the core software is sending long press information correctly to the integrations. I agree that it doesn't seem much better, if at all.

ItsHarper commented:
@zehnm This prompted me to do an experiment. Since this integration should now support long presses correctly, I compared the remote's behavior with that of a USB keyboard (the thing that's effectively being emulated). I can confirm that the keyboard behaves the way you would expect, while the remote seems to still be sending a series of individual presses.

Is the core software not sending a single pair of start and end commands to integrations when a key gets held down?

zehnm commented:
> Is the core software not sending a single pair of start and end commands to integrations when a key gets held down?

No, the press-and-hold feature is not yet ready. See #91
